diff -Nur tORIGINAL/projects/moo/src/ext-FUP_options.h d20040322/projects/moo/src/ext-FUP_options.h
--- tORIGINAL/projects/moo/src/ext-FUP_options.h	2004-03-22 11:25:05.000000000 -0500
+++ d20040322/projects/moo/src/ext-FUP_options.h	2004-03-22 11:29:37.000000000 -0500
@@ -26,10 +26,14 @@
 You can also make this base directory either read-only or read-and-write.
 It is read-only by default. To make it writable from the MOO, comment the line:
 */
-     #define EXTERN_FILES_DIR_READ_ONLY
+//     #define EXTERN_FILES_DIR_READ_ONLY
 
 
 /**********************************************************
 The default mode for new directories can be configured too:
 */
      #define CREATE_NEW_DIR_MODE		0755
+
+#define INCLUDE_FILECHMOD
+#define INCLUDE_FILERUN
+
