diff -Nur tORIGINAL/projects/moo/src/Makefile.in d20040322/projects/moo/src/Makefile.in
--- tORIGINAL/projects/moo/src/Makefile.in	2004-03-22 11:14:53.000000000 -0500
+++ d20040322/projects/moo/src/Makefile.in	2004-03-22 11:28:31.000000000 -0500
@@ -31,7 +31,7 @@
 
 CSRCS = ast.c code_gen.c db_file.c db_io.c db_objects.c db_properties.c \
 	db_verbs.c decompile.c disassemble.c eval_env.c eval_vm.c \
-	exceptions.c execute.c extensions.c functions.c keywords.c list.c \
+	exceptions.c execute.c extensions.c ext-FUP.c functions.c keywords.c list.c \
 	log.c malloc.c match.c md5.c name_lookup.c network.c net_mplex.c \
 	net_proto.c numbers.c objects.c parse_cmd.c pattern.c program.c \
 	property.c quota.c ref_count.c regexpr.c server.c storage.c streams.c str_intern.c \
