diff -Nur tORIGINAL/projects/moo/src/network.h d20040322/projects/moo/src/network.h
--- tORIGINAL/projects/moo/src/network.h	2004-03-22 11:14:54.000000000 -0500
+++ d20040322/projects/moo/src/network.h	2004-03-22 11:32:06.000000000 -0500
@@ -188,6 +188,8 @@
 				 * the given setting if valid.
 				 */
 
+extern const char *network_connection_user(network_handle nh);
+
 #ifdef OUTBOUND_NETWORK
 #include "structures.h"
 
