diff -Nur d20040322/projects/moo/src/server.c d20040323/projects/moo/src/server.c
--- d20040322/projects/moo/src/server.c	2004-03-22 13:26:05.000000000 -0500
+++ d20040323/projects/moo/src/server.c	2004-03-23 09:11:40.000000000 -0500
@@ -993,8 +993,18 @@
 void
 server_receive_line(server_handle sh, const char *line)
 {
+    int v = server_int_option("idle_time", 1800);
+    Var args;
     shandle *h = (shandle *) sh.ptr;
 
+    if ((time(0) - h->last_activity_time) > v) {
+        args = new_list(2);
+        args.v.list[1].type = TYPE_OBJ;
+        args.v.list[1].v.obj = h->player;
+        args.v.list[2].type = TYPE_INT;
+        args.v.list[2].v.num = (time(0) - h->last_activity_time);
+        run_server_task(h->player, h->listener, "user_unidled", args, "", 0);
+    }
     h->last_activity_time = time(0);
     new_input_task(h->tasks, line);
 }
