diff -Nur d20040416/projects/moo/src/extensions.c d20041022/projects/moo/src/extensions.c
--- d20040416/projects/moo/src/extensions.c	2004-04-16 16:17:24.000000000 -0400
+++ d20041022/projects/moo/src/extensions.c	2004-10-21 22:54:18.000000000 -0400
@@ -237,13 +237,19 @@
     free_var(arglist);
 
     if (nargs != 0) {
-        unsigned long newRegA = (unsigned long) *arglist.v.list[1].v.list[1].v.fnum;
-        unsigned long newRegB = (unsigned long) *arglist.v.list[1].v.list[2].v.fnum;
-        unsigned long newRegC = (unsigned long) *arglist.v.list[1].v.list[3].v.fnum;
-
-        regA = newRegA;
-        regB = newRegB;
-        regC = newRegC;
+	if (nargs != 1 || arglist.v.list[1].v.list[0].v.num != 3) {
+            return make_error_pack(E_INVARG);
+	} else if (arglist.v.list[1].v.list[1].type != TYPE_FLOAT || arglist.v.list[1].v.list[2].type != TYPE_FLOAT || arglist.v.list[1].v.list[3].type != TYPE_FLOAT) {
+            return make_error_pack(E_INVARG);
+        } else {
+            unsigned long newRegA = (unsigned long) *arglist.v.list[1].v.list[1].v.fnum;
+            unsigned long newRegB = (unsigned long) *arglist.v.list[1].v.list[2].v.fnum;
+            unsigned long newRegC = (unsigned long) *arglist.v.list[1].v.list[3].v.fnum;
+
+            regA = newRegA;
+            regB = newRegB;
+            regC = newRegC;
+        }
     }
 
     r = new_list(3);
